sirepo/auth/ldap.py

```diff
@@ -0,0 +1,102 @@
+"""LDAP authentication support
+:copyright: Copyright (c) 2022-2023 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern import pkconfig
+from pykern import pkinspect
+from pykern.pkdebug import pkdlog, pkdp
+from pykern.pkcollections import PKDict
+import ldap
+import re
+import sirepo.quest
+import sirepo.util
+
+AUTH_METHOD = "ldap"
+
+AUTH_METHOD_VISIBLE = True
+
+_cfg = None
+
+# ldap dn requires escaped special chars
+_ESCAPE_DN_MAIL = re.compile(r"([,\\#<>;\"=+])")
+
+# windows login dialogue char limit for passwords
+_MAX_ENTRY = 127
+
+#: module handle
+this_module = pkinspect.this_module()
+
+#: Well known alias for auth
+user_model = "AuthEmailUser"
+
+
+class API(sirepo.quest.API):
+    @sirepo.quest.Spec("require_cookie_sentinel")
+    def api_authLdapLogin(self):
+        def _bind(creds):
+            try:
+                ldap.initialize(_cfg.server).simple_bind_s(creds.dn, creds.password)
+            except Exception as e:
+                m = "Unable to contact LDAP server"
+                if isinstance(e, ldap.INVALID_CREDENTIALS):
+                    m = "Incorrect email or password"
+                pkdlog("{} email={}", e, creds.email)
+                return sirepo.util.SReplyExc(self.reply_ok(PKDict(form_error=m)))
+
+        def _user(email):
+            m = self.auth_db.model(user_model)
+            u = m.unchecked_search_by(unverified_email=email)
+            if not u:
+                u = m.new(unverified_email=email, user_name=email)
+                u.save()
+            return u
+
+        def _validate_entry(creds, field):
+            if not creds[field]:
+                e = "falsey"
+            elif len(creds[field]) > _MAX_ENTRY:
+                e = "over max chars"
+            else:
+                return
+            pkdlog("{} field={}; email={}", e, field, creds.email)
+            return sirepo.util.SReplyExc(
+                self.reply_ok(PKDict(form_error="Invalid user and/or password"))
+            )
+
+        req = self.parse_post()
+        res = PKDict(
+            email=req.req_data.email,
+            password=req.req_data.password,
+            dn="mail="
+            + re.sub(_ESCAPE_DN_MAIL, r"\\\1", req.req_data.email)
+            + _cfg.dn_suffix,
+        )
+        r = (
+            _validate_entry(res, "email")
+            or _validate_entry(res, "password")
+            or _bind(res)
+        )
+        if r:
+            raise r
```

Comment:
The previous review said to drop the raise for return. What was I meant was `return` from the whole api method. An api returns a reply (or a dict, but we want "ok" in this case) so code would just be:

`return self.reply_ok(PKDict(form_error=r))`